### PR TITLE
CRM-19015 Merge data loss on zero fields (batch mode)

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
- [CRM-19015: Merge data loss on zero fields (batch mode)](https://issues.civicrm.org/jira/browse/CRM-19015)

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
test this please

File: CRM/Dedupe/Merger.php [hunk 1]
```diff
@@ -862,15 +862,9 @@ public static function skipMerge($mainId, $otherId, &$migrationInfo, $mode = 'sa
...
         if (
-          (!empty($migrationInfo['rows'][$key]['main'])
-            // For custom fields a 0 (e.g in an int field) could be a true conflict. This
-            // is probably true for other fields too - e.g. 'do_not_email' but
-            // leaving that investigation as a @todo - until tests can be written.
-            // Note the handling of this has test coverage - although the data-typing
-            // of '0' feels flakey we have insurance.
-            || ($migrationInfo['rows'][$key]['main'] === '0' && substr($key, 0, 12) == 'move_custom_')
-          )
+          ($migrationInfo['rows'][$key]['main'] !== '' && $migrationInfo['rows'][$key]['main'] !== NULL)
```
Comment (JKingsnorth) [inline CRM/Dedupe/Merger.php]: Will losing )&& substr($key, 0, 12) == 'move_custom_')) have any effect? I'm just testing a few scenarios now...

**Comment by JKingsnorth** [comment]
@eileenmcnaughton I wasn't able to recreate the original problem in master. Ie: I tried merging a deceased contact with a living one and I hit a merge conflict. I tested this both ways round. Ie: with the main contact as deceased, and the other contact as deceased.

In doing some testing I did hit upon https://issues.civicrm.org/jira/browse/CRM-19074 though

**Comment by eileenmcnaughton** [comment]
hmm  - you did that through the api? I think this problem probably affects other 0 fields...

**Comment by eileenmcnaughton** [comment]
@JKingsnorth note that in batch merge it matters which of the contacts has a zero in a field the to_keep or the to_remove (presumably also applies to do_not_trade etc)

**Comment by JKingsnorth** [comment]
I didn't test through the API, I used the 'batch merge' button on the listing.

**Comment by eileenmcnaughton** [comment]
I incorporated the test into #8695  - I left out the actual change since it wasn't reproducing (even in the test!)

**Pull request closed without merging**